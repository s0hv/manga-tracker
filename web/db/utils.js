import {
  NUMERIC_VALUE_OUT_OF_RANGE,
  INVALID_TEXT_REPRESENTATION,
  UNIQUE_VIOLATION,
  IN_FAILED_SQL_TRANSACTION,
  FOREIGN_KEY_VIOLATION,
  NOT_NULL_VIOLATION,
} from 'pg-error-constants';

import { pgp } from '.';
import { NoColumnsError } from './errors.js';
import { dbLogger } from '../utils/logging.js';
import { StatusError } from '../utils/errors.js';

/**
 * Generate update statement from an object while filtering out undefined values.
 * The generated update statement lacks a WHERE clause and has the values embedded in the query.
 * @param {Object} o Input object
 * @param {String} tableName Name of the table to be updated
 * @returns {String} The update statement without a whereclause
 */
export const generateUpdate = (o, tableName) => {
  const obj = { ...o };
  Object.keys(obj).forEach(key => obj[key] === undefined && delete obj[key]);
  try {
    return pgp.helpers.update(obj, null, tableName);
  } catch (e) {
    throw new NoColumnsError('No valid columns given');
  }
};

export function handleError(err, res, msgOverrides = {}) {
  if (typeof err?.getErrors === 'function') {
    err = err.getErrors().filter(e => e?.code !== IN_FAILED_SQL_TRANSACTION)[0] || err;
  }

  if (err instanceof StatusError) {
    res.status(err.status).json({ error: err.message });
    return;
  }

  if (err instanceof NoColumnsError) {
    res.status(400).json({ error: err.message || 'No valid columns given' });
    return;
  }

  const msg = msgOverrides[err.code];
  if (err.code === INVALID_TEXT_REPRESENTATION) {
    dbLogger.debug(err.message);
    res.status(400).json({ error: msg || 'Invalid data type given' });
  } else if (err.code === NUMERIC_VALUE_OUT_OF_RANGE) {
    res.status(400).json({ error: msg || 'Number value out of range' });
  } else if (err.code === UNIQUE_VIOLATION) {
    res.status(422).json({ error: msg || 'Resource already exists' });
  } else if (err.code === FOREIGN_KEY_VIOLATION) {
    res.status(404).json({ error: msg || 'Foreign key violation' });
  } else if (err.code === NOT_NULL_VIOLATION) {
    res.status(400).json({ error: msg || 'Not null value was null' });
  } else {
    dbLogger.error(err, 'Unknown database error');
    res.status(500).json({ error: msg || 'Internal server error' });
  }
}
