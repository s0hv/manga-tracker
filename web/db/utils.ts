import {
  FOREIGN_KEY_VIOLATION,
  IN_FAILED_SQL_TRANSACTION,
  INVALID_TEXT_REPRESENTATION,
  NOT_NULL_VIOLATION,
  NUMERIC_VALUE_OUT_OF_RANGE,
  UNIQUE_VIOLATION,
} from 'pg-error-constants';
import type { Response } from 'express-serve-static-core';

import { NoColumnsError } from './errors';
import { dbLogger } from '../utils/logging.js';
import { StatusError } from '../utils/errors.js';
import type { Db } from '.';

/**
 * Generate update statement from an object while filtering out undefined values.
 * Do not pass untrusted properties to this method, as it will update every column given to it
 * @param {Object} o Input object
 * @param {Db} sql Database instance
 */
export const generateUpdate = (o: any, sql: Db) => {
  const obj = { ...o };
  Object.keys(obj).forEach(key => obj[key] === undefined && delete obj[key]);

  if (Object.keys(obj).length === 0) throw new NoColumnsError('No valid columns given');

  return sql(obj);
};

export function handleError(err: any, res: Response, msgOverrides: any = {}) {
  if (typeof err?.getErrors === 'function') {
    err = err.getErrors().filter((e: any) => e?.code !== IN_FAILED_SQL_TRANSACTION)[0] || err;
  }

  if (err instanceof StatusError) {
    res.status(err.status).json({ error: err.message });
    return;
  }

  if (err instanceof NoColumnsError) {
    res.status(400).json({ error: err.message || 'No valid columns given' });
    return;
  }

  const msg = msgOverrides[err.code];
  if (err.code === INVALID_TEXT_REPRESENTATION) {
    dbLogger.debug(err.message);
    res.status(400).json({ error: msg || 'Invalid data type given' });
  } else if (err.code === NUMERIC_VALUE_OUT_OF_RANGE) {
    res.status(400).json({ error: msg || 'Number value out of range' });
  } else if (err.code === UNIQUE_VIOLATION) {
    res.status(422).json({ error: msg || 'Resource already exists' });
  } else if (err.code === FOREIGN_KEY_VIOLATION) {
    res.status(404).json({ error: msg || 'Foreign key violation' });
  } else if (err.code === NOT_NULL_VIOLATION) {
    res.status(400).json({ error: msg || 'Not null value was null' });
  } else {
    dbLogger.error(err, 'Unknown database error');
    res.status(500).json({ error: msg || 'Internal server error' });
  }
}
