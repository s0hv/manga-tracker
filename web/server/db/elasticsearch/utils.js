export const handleElasticError = (err, res) => {
  const error = 'Internal server error';
  console.error(err);
  res.status(500).json({ error });
};

export const extractFields = (result, fields, idPrefix, customFieldFormatter) => {
  idPrefix = idPrefix ? `${idPrefix}_` : '';
  return result.body.hits.hits.map(hit => {
    const o = {
      [`${idPrefix}id`]: Number(hit._id),
      score: hit._score,
    };
    if (fields) {
      fields.forEach(field => {
        o[field] = (hit.fields[field] || [])[0];
      });

      if (customFieldFormatter) {
        const formatted = customFieldFormatter(hit.fields);
        Object.keys(formatted).forEach(k => { o[k] = formatted[k] });
      }
    }

    return o;
  });
};
