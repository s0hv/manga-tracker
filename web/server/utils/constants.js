export const csrfMissing = 'Missing CSRF token';
export const isDev = process.env.NODE_ENV !== 'production';
export const isTest = process.env.NODE_ENV === 'test';
export const NO_GROUP = 1;
