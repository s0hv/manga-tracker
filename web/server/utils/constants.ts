export const csrfMissing = 'CSRF error. Modifying requests must come from the same origin.';
export const isDev = process.env.NODE_ENV !== 'production';
export const isTest = process.env.NODE_ENV === 'test';
export const NO_GROUP = 1;

export const isSecure = (process.env.NEXTAUTH_URL || '').startsWith('https://');
export const cookiePrefix = (process.env.NEXTAUTH_URL || '').startsWith('https://') ? '__Secure-' : '';
