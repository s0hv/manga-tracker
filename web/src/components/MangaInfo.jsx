import { Typography } from '@material-ui/core';
import { makeStyles } from '@material-ui/core/styles';
import React from 'react';
import PropTypes from 'prop-types';

import {
  defaultDateDistanceToNow,
  defaultDateFormat,
  statusToString,
} from '../utils/utilities';

const useStyles = makeStyles((theme) => ({
  detailText: {
    marginLeft: '5px',
    [theme.breakpoints.down('sm')]: {
      marginLeft: '3px',
    },
  },
  infoTable: {
    width: 'max-content',
    borderSpacing: '0px 2px',
    margin: theme.spacing(1),
    '& tr:nth-child(odd)': {
      backgroundColor: theme.palette.action.hover,
    },
    '& th': {
      textAlign: 'end',
      paddingLeft: '2px',
    },
    '& td': {
      paddingRight: '4px',
    },
  },
}));


const MangaInfo = ({ mangaData, showId = false }) => {
  const latestRelease = mangaData.latestRelease ?
    new Date(mangaData.latestRelease) :
    null;
  const estimatedRelease = new Date(mangaData.estimatedRelease);


  const classes = useStyles();
  return (
    <table className={classes.infoTable} aria-label='manga information'>
      <tbody>
        {showId && (
          <tr>
            <th scope='row'><Typography>Manga id:</Typography></th>
            <td><Typography className={classes.detailText}>{mangaData.mangaId}</Typography></td>
          </tr>
        )}
        <tr>
          <th scope='row'><Typography>Latest release:</Typography></th>
          <td>
            <Typography className={classes.detailText}>
              {latestRelease ?
                defaultDateFormat(latestRelease) + ' - ' + defaultDateDistanceToNow(latestRelease) :
                'Unknown'}
            </Typography>
          </td>
        </tr>
        <tr>
          <th scope='row'><Typography>Estimated release interval:</Typography></th>
          <td>
            <Typography className={classes.detailText}>
              {(mangaData.releaseInterval ?
                `${mangaData.releaseInterval?.days || 0} days ${mangaData.releaseInterval?.hours || 0} hours` :
                'Unknown')}
            </Typography>
          </td>
        </tr>
        <tr>
          <th scope='row'><Typography>Estimated next release:</Typography></th>
          <td>
            <Typography className={classes.detailText}>
              {defaultDateFormat(estimatedRelease)}
            </Typography>
          </td>
        </tr>
        <tr>
          <th scope='row'><Typography>Latest chapter:</Typography></th>
          <td>
            <Typography className={classes.detailText}>
              {mangaData.latestChapter ? mangaData.latestChapter : 'Unknown'}
            </Typography>
          </td>
        </tr>
        <tr>
          <th scope='row'><Typography>Publication status:</Typography></th>
          <td>
            <Typography className={classes.detailText}>
              {statusToString(mangaData.status)}
            </Typography>
          </td>
        </tr>
      </tbody>
    </table>
  );
};

MangaInfo.propTypes = {
  mangaData: PropTypes.shape({
    latestRelease: PropTypes.string,
    estimatedRelease: PropTypes.string,
    releaseInterval: PropTypes.object,
    latestChapter: PropTypes.number,
    status: PropTypes.number,
  }).isRequired,
  showId: PropTypes.bool,
};

export default MangaInfo;
