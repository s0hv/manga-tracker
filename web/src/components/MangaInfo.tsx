import { Typography } from '@mui/material';
import { styled } from '@mui/material/styles';
import React, { FunctionComponent } from 'react';

import {
  defaultDateDistanceToNow,
  defaultDateFormat,
  statusToString,
} from '../utils/utilities';
import { MangaStatus, PostgresInterval } from '../../types/dbTypes';

const DetailText = styled(Typography)(({ theme }) => ({
  marginLeft: theme.spacing(0.5),
}));

const InfoTable = styled('table')(({ theme }) => ({
  width: 'max-content',
  borderSpacing: '0px 2px',
  margin: theme.spacing(1),
  '& tr:nth-of-type(odd)': {
    backgroundColor: theme.palette.action.hover,
  },
  '& th': {
    textAlign: 'end',
    paddingLeft: '2px',
  },
  '& td': {
    paddingRight: '4px',
  },
}));

export type MangaInfoProps = {
  mangaData: {
    mangaId: number,
    latestRelease?: string
    estimatedRelease?: string
    releaseInterval?: PostgresInterval
    latestChapter?: number
    status: MangaStatus
  },
  showId?: boolean,
}


const MangaInfo: FunctionComponent<MangaInfoProps> = ({ mangaData, showId = false }) => {
  const latestRelease = mangaData.latestRelease ?
    new Date(mangaData.latestRelease) :
    null;
  const estimatedRelease = mangaData.estimatedRelease ?
    new Date(mangaData.estimatedRelease) :
    null;

  return (
    <InfoTable aria-label='manga information'>
      <tbody>
        {showId && (
          <tr>
            <th scope='row'><Typography>Manga id:</Typography></th>
            <td><DetailText>{mangaData.mangaId}</DetailText></td>
          </tr>
        )}
        <tr>
          <th scope='row'><Typography>Latest release:</Typography></th>
          <td>
            <DetailText>
              {latestRelease ?
                defaultDateFormat(latestRelease) + ' - ' + defaultDateDistanceToNow(latestRelease) :
                'Unknown'}
            </DetailText>
          </td>
        </tr>
        <tr>
          <th scope='row'><Typography>Estimated release interval:</Typography></th>
          <td>
            <DetailText>
              {(mangaData.releaseInterval ?
                `${mangaData.releaseInterval?.days || 0} days ${mangaData.releaseInterval?.hours || 0} hours` :
                'Unknown')}
            </DetailText>
          </td>
        </tr>
        <tr>
          <th scope='row'><Typography>Estimated next release:</Typography></th>
          <td>
            <DetailText>
              {defaultDateFormat(estimatedRelease)}
            </DetailText>
          </td>
        </tr>
        <tr>
          <th scope='row'><Typography>Latest chapter:</Typography></th>
          <td>
            <DetailText>
              {mangaData.latestChapter ? mangaData.latestChapter : 'Unknown'}
            </DetailText>
          </td>
        </tr>
        <tr>
          <th scope='row'><Typography>Publication status:</Typography></th>
          <td>
            <DetailText>
              {statusToString(mangaData.status)}
            </DetailText>
          </td>
        </tr>
      </tbody>
    </InfoTable>
  );
};

export default MangaInfo;
