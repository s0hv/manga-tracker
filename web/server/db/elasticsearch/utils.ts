import { ApiResponse } from '@elastic/elasticsearch';
import type { Response } from 'express-serve-static-core';

import type {
  MangaSearchResult,
  MangaSearchResultFields,
} from '@/db/elasticsearch/manga';

export const handleElasticError = (err: unknown, res: Response) => {
  const error = 'Internal server error';
  console.error(err);
  res.status(500).json({ error });
};

export type ExtractedFields = {
  score: number
  [key: string]: unknown
};
export type CustomFieldFormatter<T extends boolean> = (fields: MangaSearchResultFields<T>) => Record<string, unknown>;

export const extractFields = <T extends boolean>(
  result: ApiResponse<MangaSearchResult<T>>,
  fields?: (keyof MangaSearchResultFields<T>)[],
  idPrefix?: string,
  customFieldFormatter?: (fields: MangaSearchResultFields<T>) => Record<string, unknown>
) => {
  const mappedIdPrefix = idPrefix ? `${idPrefix}_` : '';
  return result.body.hits.hits.map(hit => {
    const o: ExtractedFields = {
      [`${mappedIdPrefix}id`]: Number(hit._id),
      score: hit._score,
    };
    if (fields) {
      fields.forEach(field => {
        o[field as string] = (hit.fields[field] as Array<unknown> ?? [])[0];
      });

      if (customFieldFormatter) {
        const formatted = customFieldFormatter(hit.fields);
        Object.keys(formatted).forEach(k => {
          o[k] = formatted[k];
        });
      }
    }

    return o;
  });
};
