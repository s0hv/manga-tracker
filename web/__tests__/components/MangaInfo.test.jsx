import React from 'react';
import { render, screen } from '@testing-library/react';

import MangaInfo from '../../src/components/MangaInfo';
import { mockUTCDates } from '../utils';
import {
  defaultDateFormatRegex,
  defaultDateDistanceFormat
} from '../constants';

describe('MangaInfo should render correctly', () => {
  mockUTCDates();
  const mangaData = {
    mangaId: 1,
    releaseInterval: {
      days: 7,
    },
    latestRelease: '2020-07-05T16:00:00.000Z',
    estimatedRelease: '2020-07-12T16:00:00.000Z',
    latestChapter: 157,
    status: 0,
  };

  const nullData = {
    mangaId: 1,
    status: 0,
  };

  it('Should render correctly with data', () => {
    render(<MangaInfo mangaData={mangaData} />);

    expect(screen.queryByRole('row', { name: /manga id: \d+/i })).not.toBeInTheDocument();

    expect(
      screen.getByRole(
        'row',
        { name: new RegExp(`latest release: ${defaultDateFormatRegex} - ${defaultDateDistanceFormat}`, 'i') }
      )
    ).toBeInTheDocument();

    expect(
      screen.getByRole('row', { name: /estimated release interval: \d+ days \d+ hours/i })
    ).toBeInTheDocument();

    expect(
      screen.getByRole(
        'row',
        { name: new RegExp(`estimated next release: ${defaultDateFormatRegex}`, 'i') }
      )
    ).toBeInTheDocument();

    expect(screen.getByRole('row', { name: /latest chapter: \d+/i })).toBeInTheDocument();
    expect(screen.getByRole('row', { name: /publication status: \w+/i })).toBeInTheDocument();
  });

  it('Should render correctly with null data', () => {
    render(<MangaInfo mangaData={nullData} showId />);

    expect(screen.queryByRole('row', { name: /manga id: \d+/i })).toBeInTheDocument();

    expect(
      screen.getByRole('row', { name: /latest release: unknown/i })
    ).toBeInTheDocument();

    expect(
      screen.getByRole('row', { name: /estimated release interval: unknown/i })
    ).toBeInTheDocument();

    expect(
      screen.getByRole('row', { name: /estimated next release: unknown/i })
    ).toBeInTheDocument();

    expect(screen.getByRole('row', { name: /latest chapter: unknown/i })).toBeInTheDocument();
    expect(screen.getByRole('row', { name: /publication status: \w+/i })).toBeInTheDocument();
  });

  it('Should throw TypeError when mangaData not given', () => {
    expect(() => render(<MangaInfo />)).toThrow(TypeError);
  });
});
