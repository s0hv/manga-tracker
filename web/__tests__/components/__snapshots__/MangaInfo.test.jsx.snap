// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MangaInfo should render correctly Should render correctly with data 1`] = `
<table
  aria-label="manga information"
  className="makeStyles-infoTable-2"
>
  <tbody>
    <tr>
      <th
        scope="row"
      >
        <p
          className="MuiTypography-root MuiTypography-body1"
        >
          Latest release:
        </p>
      </th>
      <td>
        <p
          aria-describedby={null}
          className="MuiTypography-root makeStyles-detailText-1 MuiTypography-body1"
          onBlur={[Function]}
          onFocus={[Function]}
          onMouseLeave={[Function]}
          onMouseOver={[Function]}
          onTouchEnd={[Function]}
          onTouchStart={[Function]}
          title="Sun, 05 Jul 2020 16:00:00 GMT"
        >
          Jul 5th 2020, 16:00 - 1 day ago
        </p>
      </td>
    </tr>
    <tr>
      <th
        scope="row"
      >
        <p
          className="MuiTypography-root MuiTypography-body1"
        >
          Estimated release interval:
        </p>
      </th>
      <td>
        <p
          className="MuiTypography-root makeStyles-detailText-1 MuiTypography-body1"
        >
          7 days 0 hours
        </p>
      </td>
    </tr>
    <tr>
      <th
        scope="row"
      >
        <p
          className="MuiTypography-root MuiTypography-body1"
        >
          Estimated next release:
        </p>
      </th>
      <td>
        <p
          className="MuiTypography-root makeStyles-detailText-1 MuiTypography-body1"
        >
          Jul 12th 2020, 16:00
        </p>
      </td>
    </tr>
    <tr>
      <th
        scope="row"
      >
        <p
          className="MuiTypography-root MuiTypography-body1"
        >
          Latest chapter:
        </p>
      </th>
      <td>
        <p
          className="MuiTypography-root makeStyles-detailText-1 MuiTypography-body1"
        >
          157
        </p>
      </td>
    </tr>
    <tr>
      <th
        scope="row"
      >
        <p
          className="MuiTypography-root MuiTypography-body1"
        >
          Publication status:
        </p>
      </th>
      <td>
        <p
          className="MuiTypography-root makeStyles-detailText-1 MuiTypography-body1"
        >
          Ongoing
        </p>
      </td>
    </tr>
  </tbody>
</table>
`;
